### PR TITLE
[RF] Replaced legacy iterators with range-based loops

### LINKED ISSUE #8777: [RF] Migration from RooAbsCollection and RooLinkedList legacy iterators to range-based loops
**This migration is a great first issue for an enthusiastic open source contributor who wants to get started contributing to ROOT with a well-defined task!**

**This issue has already been addressed partially by many PRs, the only remaining iterator-creating function to look out for and replace is **`RooLinkedList::MakeIterator()`**!**

## Introduction

RooFit needs to be migrated to use range-based loops to iterate over RooArgLists, RooArgSets, and RooLinkedLists. The [deprecated iterator-creating functions](https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooAbsCollection.h#L178) `RooAbsCollection::createIterator()`, `RooAbsCollection::iterator()`, and `RooAbsCollection::fwdIterator()` have some overhead both is code verbosity and performance. For the RooLinkedList, the relevant function calls to replace are `RooLinkedList::MakeIterator()`, `RooLinkedList::iterator()`, and `RooLinkedList::fwdIterator()`.

To do the migration, one could remove the deprecated functions from the code and try to recompile ROOT. From the compiler errors, one can see where a migration away from the legacy iterators is needed. Alternatively, one could mark the legacy functions as `[[deprecated]]` and get compiler warnings.

If the iteration involved a type casting of each item, one can also use the `static_range_cast` and `dynamic_range_cast` functions in RooFit to make the code more compact (introduced in https://github.com/root-project/root/pull/8769).

It is not expected that the full migration is done in one pull request, as these legacy iterations appear everywhere in RooFit. One can proceed with occasional PRs, migrating only one or a few RooFit classes at the time.

## Migration examples

Here are some examples on how the migration to range-based loops can look like

1. Basic example:
    ```C++
    // with legacy iterator
    TIterator* iter = comps->createIterator() ;
    while((arg=(RooAbsArg*)iter->Next())) { 
      // do something with arg
    }
   ```

   ```C++
    // with range-based loop
    for(auto const* arg : *comps) { 
      // do something with arg
    }
    ```

2. Example where one can make use of `static_dynamic_cast`
    ```C++
    // with legacy iterator
    RooFIter siter = fullPdfSet.fwdIterator() ;
    RooAbsPdf* pdf ;
    while((pdf=(RooAbsPdf*)siter.next())) {
      // do something with pdf
    }
   ```

   ```C++
    // with range-based loop
    for(auto const* pdf : static_range_cast<RooAbsPdf*>(fullPdfSet)) {
      // do something with pdf
    }
    ```

3. Example where one can make use of `static_dynamic_cast`
    ```C++
    // with legacy iterator
    TIterator* uIter = usedObs->createIterator() ;
    RooAbsArg* obs ;
    while((obs=(RooAbsArg*)uIter->Next())) {
      RooRealVar* rrv = dynamic_cast<RooRealVar*>(obs) ;
      if (rrv) {
        // ... do something with rrv
      }
    }
   ```

   ```C++
    // with range-based loop
    for(auto const* rrv : dynamic_range_cast<RooRealVar*>(*usedObs)) {
      if (rrv) {
        // ... do something with rrv
      }
    }
    ```

### PR DESCRIPTION
**Pull request by HannaOlvhammar**
# This Pull request:

## Changes or fixes:
Replaced while loops with the legacy iterators fwdIterator() with range based for loops. 

## Checklist:

- [X] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #8777

### CODE REVIEW

File: roofit/roofitcore/src/RooAbsAnaConvPdf.cxx [hunk 1]
```diff
@@ -151,10 +151,7 @@ RooAbsAnaConvPdf::~RooAbsAnaConvPdf()
 ///
 /// The return value is a unique identifier code, that will be passed to coefficient()
 /// to identify the basis function for which the coefficient is requested. If the
-/// resolution model used does not support the declared basis function, code -1 is
-/// returned.
-///
-
+/// resolution model used dsource root_install/bin/thisroot.sh
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsAnaConvPdf.cxx:154]: Something got messed up here in the documentation, can you please revert that?

File: roofit/roofitcore/src/RooAbsTestStatistic.cxx [hunk 2]
```diff
@@ -507,9 +507,7 @@ void RooAbsTestStatistic::initSimMode(RooSimultaneous* simpdf, RooAbsData* data,
         binnedL = true ;
       } else if (pdf->IsA()->InheritsFrom(RooProdPdf::Class())) {
         // Default case: top-level pdf is a product of RRSP and other pdfs
-        RooFIter iter = ((RooProdPdf*)pdf)->pdfList().fwdIterator() ;
-        RooAbsArg* component ;
-        while ((component = iter.next())) {
+        for (auto const* component : ((RooProdPdf*)pdf)->pdfList()) {
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsTestStatistic.cxx:510]:
```suggestion
        for (auto const* component : static_cast<RooProdPdf*>(pdf)->pdfList()) {
```
When you are already touching the code, you should replace C-style casts with C++ casts please

File: roofit/roofitcore/src/RooAddModel.cxx [hunk 3]
```diff
@@ -110,10 +110,10 @@ RooAddModel::RooAddModel(const char *name, const char *title, const RooArgList&
    }
 
    // Constructor with N PDFs and N or N-1 coefs
-   auto pdfIter = inPdfList.fwdIterator();
-
+   std::size_t i = 0;
    for (auto const &coef : inCoefList) {
-      auto pdf = pdfIter.next();
+      //auto pdf = pdfIter.next();
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooAddModel.cxx:115]:
```suggestion
```
This commented-out line can be deleted

File: roofit/roofitcore/src/RooLinearVar.cxx [hunk 4]
```diff
@@ -248,9 +248,7 @@ std::list<std::string> RooLinearVar::getBinningNames() const
 {
   std::list<std::string> binningNames(1, "");
 
-  RooFIter iter = _altBinning.fwdIterator();
-  const RooAbsArg* binning = 0;
-  while((binning = iter.next())) {
+  for (auto const* binning : static_range_cast<RooAbsArg*>(_altBinning)) {
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooLinearVar.cxx:251]:
```suggestion
  for (RooAbsArg const* binning : _altBinning) {
```
No need to cast here, the class is already RooAbsArg (maybe replace the `auto` with RooAbsArg too such that it's explicitly clear from now).

File: roofit/roofitcore/src/RooProdPdf.cxx [hunk 5]
```diff
@@ -1189,9 +1184,7 @@ void RooProdPdf::rearrangeProduct(RooProdPdf::CacheElem& cache) const
 
 //    cout << "denominator components are " << dentmp->components() << endl ;
    RooArgSet comps(dentmp->components()) ;
-   RooFIter piter = comps.fwdIterator() ;
-   RooAbsReal* parg ;
-   while((parg=(RooAbsReal*)piter.next())) {
+   for (auto* parg : static_range_cast<RooAbsReal*>(comps)) {
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooProdPdf.cxx:1187]:
```suggestion
   for (auto* parg : static_range_cast<RooAbsReal*>(dentmp->components())) {
```

I don't this that `comps` is used anywhere else, so you can delete the precious line and avoid creating a RooArgSet by copy here I think.

File: roofit/roofitcore/src/RooProduct.cxx [hunk 6]
```diff
@@ -126,11 +126,11 @@ bool RooProduct::forceAnalyticalInt(const RooAbsArg& dep) const
...
   // of the product terms depend on it.
 
-  RooFIter compRIter = _compRSet.fwdIterator() ;
-  RooAbsReal* rcomp ;
   bool depends(false);
-  while((rcomp=(RooAbsReal*)compRIter.next())&&!depends) {
+  for (auto const* rcomp : static_range_cast<RooAbsReal*>(_compRSet)) {
+    if (!depends) {
         depends = rcomp->dependsOn(dep);
+    }
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooProduct.cxx:133]:
```suggestion
  for (auto const* rcomp : static_range_cast<RooAbsReal*>(_compRSet)) {
    if (depends) break;
    depends = rcomp->dependsOn(dep);
```
The logic is not the same now. Previously, the look was stopped if `depends` is true, and now you just skip one iteration. You need to do something different, for example use `break`.

**Comment by guitargeek** [comment]
@phsft-bot build

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi Hanna, thank you very much for this PR! I have only a few comments, please address then and then we can merge this and finally close the GitHub issue!

**Comment by hageboeck** [comment]
PS: If reasonably possible, *every single commit* should compile. If you make a mistake and want to fix it in a second commit, no problem. Git has `git commit --fixup <commitHash>` + interactive rebase for this situation.

**Comment by guitargeek** [comment]
@phsft-bot build

File: roofit/roofitcore/src/RooLinearVar.cxx [hunk 7]
```diff
@@ -248,9 +248,7 @@ std::list<std::string> RooLinearVar::getBinningNames() const
 {
   std::list<std::string> binningNames(1, "");
 
-  RooFIter iter = _altBinning.fwdIterator();
-  const RooAbsArg* binning = 0;
-  while((binning = iter.next())) {
+  for (RooAbsArg const* binning : _altBinning) {
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooLinearVar.cxx:251]:
```suggestion
  for (TObject const* binning : _altBinning) {
```
The code doesn't compile now, as the type is not `RooAbsArg` but `TObject`. I think just changing this should fix it (sorry for suggesting the wrong type in the previous comment!).

**Comment by guitargeek** [comment]
@phsft-bot build

**Review by guitargeek — APPROVED** [review]
LGTM, thank you very much @HannaOlvhammar for this!

**Pull request merged**